File: app/src/main/java/com/xenris/liquidwarsos/server/ServerFinder.java


package com.xenris.liquidwarsos.server;

import android.content.Context;

import java.net.InetAddress;
import java.net.SocketException;
import java.net.DatagramSocket;
import java.net.DatagramPacket;
import java.io.IOException;
import java.lang.Thread;

public class ServerFinder {
    private static DatagramSocket datagramSocket = null;
    private static final String queryMessage = "9287349213846";
    private static final String responseMessage = "2138479287348";
    private static boolean isSearching = false;

    public static void share(final Context context, final int port, final String serverName) {
        if(datagramSocket != null)
            return;
        new Thread() {
            @Override
            public void run() {
                try {
                    datagramSocket = new DatagramSocket(port);
                    byte[] receiveData = new byte[32];
                    byte[] sendData = (responseMessage + serverName).getBytes();
                    String broadcastAddress = NetInfo.getBroadcastAddress(context);
                    DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
                    while(true) {
                        datagramSocket.receive(receivePacket);
                        String s = new String(receivePacket.getData(), 0, receivePacket.getLength());
                        if(s.compareTo(queryMessage) == 0) {
                            InetAddress inIP = receivePacket.getAddress();
                            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, inIP, receivePacket.getPort());
                            datagramSocket.send(sendPacket);
                        }
                    }
                } catch (SocketException s) { } catch (IOException i) { }
            }
        }.start();
    }

    public static void stopSharing() {
        if(datagramSocket != null)
            datagramSocket.close();
        datagramSocket = null;
    }

    public static void search(final ServerFinderCallbacks sfc, final String ip, final int port) {
        if(isSearching)
            return;
        new Thread() {
            @Override
            public void run() {
                isSearching = true;
                try {
                    datagramSocket = new DatagramSocket(port);
                    datagramSocket.setSoTimeout(1000);
                    byte[] receiveData = new byte[128];
                    byte[] sendData = queryMessage.getBytes();
                    DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, InetAddress.getByName(ip), port);
                    DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
                    while(isSearching) {
                        datagramSocket.send(sendPacket);
                        while(true) {
                            try {
                                datagramSocket.receive(receivePacket);
                            } catch (IOException i) {
                                break;
                            }
                            byte[] data = receivePacket.getData();
                            int l = responseMessage.getBytes().length;
                            String s = new String(data, 0, l);
                            if(s.compareTo(responseMessage) == 0) {
                                String name = new String(data, l, receivePacket.getLength() - l);
                                String address = receivePacket.getAddress().getHostAddress();
                                sfc.onServerFound(new ServerInfo(name, address));
                            }
                        }
                    }
                } catch (SocketException s) {
                } catch (IOException i) {
                }
                stopSearching();
            }
        }.start();
    }

    public static class ServerInfo {
        public String name;
        public String ip;

        public ServerInfo(String n, String i) {
            name = n;
            ip = i;
        }
    }

    public static void stopSearching() {
        if(datagramSocket != null)
            datagramSocket.close();
        datagramSocket = null;
        isSearching = false;
    }

    public interface ServerFinderCallbacks {
        public void onServerFound(ServerInfo serverInfo);
    }
}

File: app/src/main/java/com/dergoogler/liquidwars/server/ServerFinder.kt
package com.dergoogler.liquidwars.server

import android.content.Context
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

object ServerFinder {
    private var datagramSocket: DatagramSocket? = null
    private const val queryMessage = "9287349213846"
    private const val responseMessage = "2138479287348"
    private var isSearching = false

    fun share(context: Context, port: Int, serverName: String) {
        if (datagramSocket != null) return
        object : Thread() {
            override fun run() {
                try {
                    datagramSocket = DatagramSocket(port)
                    val receiveData = ByteArray(32)
                    val sendData = (responseMessage + serverName).toByteArray()
                    val broadcastAddress = NetInfo.getBroadcastAddress(context)
                    val receivePacket = DatagramPacket(receiveData, receiveData.size)
                    while (true) {
                        datagramSocket!!.receive(receivePacket)
                        val s = String(receivePacket.data, 0, receivePacket.length)
                        if (s.compareTo(queryMessage) == 0) {
                            val inIP = receivePacket.address
                            val sendPacket =
                                DatagramPacket(sendData, sendData.size, inIP, receivePacket.port)
                            datagramSocket!!.send(sendPacket)
                        }
                    }
                } catch (ignored: IOException) {
                }
            }
        }.start()
    }

    fun stopSharing() {
        if (datagramSocket != null) datagramSocket!!.close()
        datagramSocket = null
    }

    fun search(sfc: ServerFinderCallbacks, ip: String?, port: Int) {
        if (isSearching) return
        object : Thread() {
            override fun run() {
                isSearching = true
                try {
                    datagramSocket = DatagramSocket(port)
                    datagramSocket!!.soTimeout = 1000
                    val receiveData = ByteArray(128)
                    val sendData = queryMessage.toByteArray()
                    val sendPacket =
                        DatagramPacket(sendData, sendData.size, InetAddress.getByName(ip), port)
                    val receivePacket = DatagramPacket(receiveData, receiveData.size)
                    while (isSearching) {
                        datagramSocket!!.send(sendPacket)
                        while (true) {
                            try {
                                datagramSocket!!.receive(receivePacket)
                            } catch (i: IOException) {
                                break
                            }
                            val data = receivePacket.data
                            val l = responseMessage.toByteArray().size
                            val s = String(data, 0, l)
                            if (s.compareTo(responseMessage) == 0) {
                                val name = String(data, l, receivePacket.length - l)
                                val address = receivePacket.address.hostAddress
                                sfc.onServerFound(ServerInfo(name, address))
                            }
                        }
                    }
                } catch (ignored: IOException) {
                }
                stopSearching()
            }
        }.start()
    }

    fun stopSearching() {
        if (datagramSocket != null) datagramSocket!!.close()
        datagramSocket = null
        isSearching = false
    }

    class ServerInfo(var name: String, var ip: String)

    interface ServerFinderCallbacks {
        fun onServerFound(serverInfo: ServerInfo?)
    }
}
